### ISSUE TITLE
Random Seeds

### ISSUE THREAD

**Issue by garrettwrong** [ComputationalCryoEM/ASPIRE-Python#209]
We have several randomized methods/calls in ASPIRE.  We also specify seeds throughout the code/test, most often an initial argument to constructors. Occasionally code is missing that detail, which would make it non deterministic for testing purposes.

I am proposing that for ASPIRE, we construct one canonical seed that is set in a single place, probably `config.ini`.  Invoking codes should reference that when no seed is provided, and generally speaking anything random should always refer to a seed.  If a method (say in numpy) does not admit a seed directly, we should figure out how the parent package can be set (ie `numpy.random.seed`) on package import.

Whether we should default to a production(0) or fixed value depends, and is probably less important at this time.

**Comment by janden** [comment]
I think that makes sense. There are a few cases where we have to be a little careful about resetting seeds between calls (think Monte Carlo stuff) since that would destroy any independence between samples. Generally, this should be quite rare and we should be able to spot and handle those issues locally.

**Comment by garrettwrong** [comment]
Yeah, I should have described a single python module for this in addition to the `confg.ini`, reading what I wrote, I missed that bit, though that is what I was thinking :).  Correct, this seed resetting has already come up, today in fact.

**Comment by junchaoxia** [comment]
It is good to have the option to set random seed explicitly. I want to add some things based on the current Python code. The need of random seed and random number generator has two major purposes: 1) The implemented algorithms or submodules need them but the detailed sequences are not important. In this case we can use any seed and any functions from the Python libs.  2) We need to reproduce the same sequences as Matlab so that we can compare the results numerically between Python and Matlab. For this case we have implemented the `randi` and `randn` in the `matlab_compat.py` and used in many places such as random rotation angles and CTF indices in ImageSource class. One of important concerns is how we should balance these two purposes and put them in a united way.

**Comment by vineetbansal** [comment]
Just making a note here that the following place in the code may need investigation:
https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/60f4ced586c308e3947489fd04279604c5ea364f/src/aspire/commands/aspire.py#L20

**Comment by janden** [comment]
Another thing to keep in mind is that the standard way of getting random numbers in NumPy (`np.random.rand*`) is frozen and will not be receiving any updates. Instead, they encourage you to use the `np.random.Generator` framework. The drawback there is that the random numbers are no longer guaranteed to stay the same across versions. For more info, see [here](https://docs.scipy.org/doc/numpy-1.17.0/reference/random/index.html).

**Comment by garrettwrong** [comment]
Some initial cleanup was done in #310 , but it doesn't completely cover the later points of having a single point of control and moving to Generator streams.  That PR simply does some of the prep work for that.  Leaving this open for now.

**Comment by garrettwrong** [comment]
Marking this as done wrt release prep (covered in #310), but leaving open because there is a good future issue here... namely migrated from the legacy numpy to Generator etc.
